File: MyBudgetBackend/src/main/java/org/unibl/etf/mybudgetbackend/services/TransactionService.java
package org.unibl.etf.mybudgetbackend.services;


import org.unibl.etf.mybudgetbackend.models.dto.TransactionDTO;
import org.unibl.etf.mybudgetbackend.models.dto.TransactionRequestDTO;

import java.util.List;

/**
 * Service interface for managing transaction-related operations.
 * Provides abstract methods for CRUD operations on transaction entities.
 */
public interface TransactionService {
    List<TransactionDTO> findAll();
    TransactionDTO findById(Long id);
    List<TransactionDTO> findAllByAccountId(Long id);
    TransactionDTO insert(Long accountId,TransactionRequestDTO request);
}

File: my-budget-frontend/src/app/features/transaction/services/transaction.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { environment } from '../../../environments/environments';
import { first, map, Observable, tap } from 'rxjs';
import {
  TransactionDTO,
  TransactionRequestDTO,
} from '../../../models/transactions';
import { MessageService } from 'primeng/api';
import { AccountService } from '../../account/services/account.service';

@Injectable({
  providedIn: 'root',
})
export class TransactionService {
  http = inject(HttpClient);
  messageService = inject(MessageService);
  accountService = inject(AccountService);

  findAll() {
    return this.http.get(`${environment.baseUrl}/transactions`).pipe(
      first(),
      map((result: any) => {
        if (!result) return [];
        return result;
      })
    );
  }

  findAllByAccountId(id: number) {
    return this.http
      .get(`${environment.baseUrl}/accounts/${id}/transactions`)
      .pipe(
        first(),
        map((result: any) => {
          if (!result) return [];
          return result;
        })
      );
  }
  insert(accountId: number, request: TransactionRequestDTO): Observable<any> {
    return this.http
      .post<TransactionDTO>(
        `${environment.baseUrl}/accounts/${accountId}/transactions`,
        request
      )
      .pipe(
        first(),
        tap((data) => {
          this.accountService.update(data.account);
          this.messageService.add({
            severity: 'success',
            summary: 'Transaction successfully created.',
          });
        })
      );
  }
}
